### PR TITLE
util: generalize accounting of system-allocated memory in pool resource

### PR DESCRIPTION
**Pull request by LarryRuane**
Follow-on to PR #25325, "Add pool based memory resource"

The `DynamicUsage()` function for the version of `unordered_map` that uses the `PoolAllocator` returns a close approximation of the amount of physical memory used by the map. (This is the map used for the dbcache.) It accounts for several of the allocator's internal data structures, such as the memory chunks and the freelist. It also includes the `unordered_map`'s bucket array (vector), which is a bit out of place because the rest of the pool allocator doesn't know or make any assumption about the type of container that the pool allocator is being used for, namely, currently only `std::unordered_map`.

This change could prevent a possible future pool memory usage calculation error, although it would likely be a small error: If the pool is configured with a large `MAX_BLOCK_SIZE_BYTES`, it could turn out that the size of the bucket array is small enough to be allocated by the resource allocator; it would then be double-counted.

Another aspect of the pool allocator that this PR improves is a refactor of the pool allocator's `DynamicUsage()` implementation, to move details about the pool allocator's internal data structures out of `memusage.h`, which is a general source file, to the allocator's `pool.h`.

### CODE REVIEW

File: src/support/allocators/pool.h [hunk 1]
```diff
@@ -164,6 +164,26 @@ class PoolResource final
...
+        if (alloc == 0) {
+            return 0;
+        } else if (sizeof(void*) == 8) {
+            return ((alloc + 31) >> 4) << 4;
+        } else if (sizeof(void*) == 4) {
+            return ((alloc + 15) >> 3) << 3;
+        } else {
+            assert(0);
+        }
+    }
```
Comment (LarryRuane) [inline src/support/allocators/pool.h:185]: Copied this from https://github.com/bitcoin/bitcoin/blob/a13f3746dccd9c4ec16d6bfe9b33ebd26e3238e1/src/memusage.h#L50 because including that file would create a circular dependency. Is there a better way to do this?

File: src/test/pool_tests.cpp [hunk 2]
```diff
@@ -54,33 +54,40 @@ BOOST_AUTO_TEST_CASE(basic_allocating)
     PoolResourceTester::CheckAllDataAccountedFor(resource);
     BOOST_TEST(1 == PoolResourceTester::FreeListSizes(resource)[1]);
     BOOST_TEST(expected_bytes_available == PoolResourceTester::AvailableMemoryFromChunk(resource));
+    // the system allocates 32 bytes for an 8-byte allocation
+    BOOST_TEST(32 == resource.SystemBytes());
```
Comment (LarryRuane) [inline src/test/pool_tests.cpp:58]: This test may be somewhat fragile; it depends on `MallocUsage(8) == 32`. Maybe the test condition here should be `8 <= resource.SystemBytes()`, but I think that doesn't test enough.

**Comment by LarryRuane** [comment]
How much of a difference does this make? I ran IBD on Ubuntu using the default `dbcache` (450) and `maxmempool` (300), which gives a total cache size (during IBD) of about 750 MB, and this (missing) system allocation amount was 97 MB (96941536 bytes).

With this amount of allocation omitted from the memory usage estimate (that is, current master), nodes will use 97 MB more memory than expected, or about 13% more than configured, a pretty significant error. This regression could cause surprises on low-memory systems.

I also verified that the operating system (Ubuntu) reports the process using more memory on master than with this PR.
```
$ grep VmData /proc/$(pgrep bitcoind)/status
```
That reports 1281764 kB on master and 1203800 on this PR branch. (This is after IBD had been running long enough to do at least one cache flush.) I'm not sure how reliable that is, or if that's the correct value to look at (`proc/status` show many memory-related values).

The `top` command reported the RES (resident) memory size as "1.7g" on master and "1.3g" on the PR branch, but that seems to be too much of a difference (about 400 MB), so take that with a grain of salt.

It would be helpful if others could try to reproduce these results.

cc @martinus @theStack @sipa @jonatack

File: src/test/pool_tests.cpp (hunk 2 above)
Comment (LarryRuane) [reply src/test/pool_tests.cpp]: CI failed here (on 32-bit), so I force-pushed to make the test more precise.

File: src/memusage.h [hunk 3]
```diff
@@ -183,7 +183,9 @@ static inline size_t DynamicUsage(const std::unordered_map<Key,
     size_t estimated_list_node_size = MallocUsage(sizeof(void*) * 3);
     size_t usage_resource = estimated_list_node_size * pool_resource->NumAllocatedChunks();
     size_t usage_chunks = MallocUsage(pool_resource->ChunkSizeBytes()) * pool_resource->NumAllocatedChunks();
-    return usage_resource + usage_chunks + MallocUsage(sizeof(void*) * m.bucket_count());
+    size_t usage_system = pool_resource->SystemBytes();
```
Comment (martinus) [inline src/memusage.h:186]: I don't think that calculation is now more correct, in fact this now counts the the index array twice; once in `usage_system` and once in the estimated `usage_buckets`.

The most correct way is probably to just let the pool do all the malloc & free accounting, and don't do any other estimation. Then the `DynamicUsage` method boils down to something like `return m.get_allocator().resource().memory_usage();`. That would give more theoretically correct numbers, but in practice I doubt that this is much different from what we already have.

I think the current inaccuray comes from either the mystery behind glibc & how the OS kernel does its thing, or some other data structures are not correctly accounted for.

Comment (LarryRuane) [reply src/memusage.h:186]: I see, yes, it's counting the buckets twice, thanks. I agree that either way gives the same answer; the way I'm suggesting here would isolate this pool code a little more from the details of `std::unordered_map` (i.e., the existence of the `m.bucket_count()` method), which would be a little more general, but I think the way it is now is fine. I guess if we use this pool resource for a different kind of container (for example, an ordered map), then this PR's approach would be better since we wouldn't have to understand the internals of, say, `std::map`.

I'll go ahead and close this now, we can keep it in mind for possible future use.

**Comment by LarryRuane** [comment]
Closing at least for now, see the [comment](https://github.com/bitcoin/bitcoin/pull/27748#discussion_r1205009503) above.

**Comment by martinus** [comment]
Also related: #26614

File: src/memusage.h (hunk 3 above)
Comment (martinus) [reply src/memusage.h:186]: I agree, also with current memory accounting code its not possible to use the same memory resource for multiple containers. So it would be nice to add full accounting to the pool, and then remove the special handling of std::unordered_map with pool and just count the ressource's total usage.

**Comment by LarryRuane** [comment]
I decided to reopen this; I think it's a good generalization improvement to the pool allocator. See the description for details. Thanks, @martinus, for the suggested improvement!

**Comment by LarryRuane** [comment]
GitHub isn't allowing me to reopen this; @achow101 could you re-open this for me, please? Or, if you'd prefer, I could make a new PR. Thanks!

TL;DR on this PR's history: When I created this PR, I thought I was fixing a bug. But @martinus pointed out that there is no bug, so I closed this PR. I then decided this would be a worthwhile improvement after all (although of course not trying to fix a bug), so I'd like it to be reopened.

**Comment by ryanofsky** [comment]
> GitHub isn't allowing me to reopen this

In order to reopen the PR, the branch needs to point at the same commit it was pointing to at the time the PR was closed.

Right now the branch https://github.com/LarryRuane/bitcoin/commits/2023-05-resource-pool-system-alloc points to d25b54346fed931830cf3f538b96c5c346165487, but the PR points to 576a270bbcbb21725770b07ba32af8e0c8965f44. If you temporarily force push 576a270bbcbb21725770b07ba32af8e0c8965f44 to the branch, I think you should be able to reopen the PR.

**Comment by LarryRuane** [comment]
Thanks, Ryan!

File: src/support/allocators/pool.h [hunk 4]
```diff
@@ -164,6 +174,21 @@ class PoolResource final
         m_allocated_chunks.emplace_back(m_available_memory_it);
     }
 
+    // copied from src/memusage.h
+    size_t MallocUsage(size_t alloc) const
+    {
+        // Measured on libc6 2.19 on Linux.
```
Comment (fanquake) [inline src/support/allocators/pool.h:180]: Not sure what value this comment adds. Also don't see a reason to mention a version of glibc that we don't support?

Comment (LarryRuane) [reply src/support/allocators/pool.h:180]: I copied this function from `memusage.h` ([here](https://github.com/bitcoin/bitcoin/blob/6cf47a8f44f96099a84e5c6e628e3499045e024d/src/memusage.h#L52)). I'm hoping not to duplicate this function, even though it's pretty small. `pool.h` can't include `memusage.h` because that creates a circular dependency. I'll try to figure out a way to improve this and also of course I welcome suggestions.

Anyway, if we need to duplicate this function, I'll remove this comment, I agree it's not useful.

Comment (martinus) [reply src/support/allocators/pool.h:180]: About that function, I do not think that it is accurate any more. I wrote a different version of that estimation that makes use of `malloc_usable_size`, and that way I get different values, see here: https://godbolt.org/z/3a6E57dPx

It might be worth having a look if using that estimation is more accurate

Comment (LarryRuane) [reply src/support/allocators/pool.h:180]: Thanks, I got a slightly different result -- `malloc_usable_size() + sizeof(void*)` is perfect (note adding only `sizeof(void*)` instead of 2 times that). I'm not set up for 32-bit, can you try running this test program on 32-bit if you are able to easily?

Whichever is the most accurate, would a PR to improve [`memusage.h: MallocUsage()`](https://github.com/bitcoin/bitcoin/blob/6cf47a8f44f96099a84e5c6e628e3499045e024d/src/memusage.h#L50) be worth doing? (I would say a separate PR would be better than doing that here; then this PR could be based on that one.)

<details>
<summary>malloc space usage estimation test program</summary>

```
// malloc space usage estimation
//
// g++ -std=c++17 -m64 t.cpp
//
// See:
// https://github.com/bitcoin/bitcoin/pull/27748#discussion_r1209810958
//
// This program does a bunch of memory allocations of various
// sizes, and calculates the memory usage increase using /proc.
//
// For each allocation size, it prints:
//   allocation size
//   malloc_usable_size() + sizeof(void*)
//   empirical estimate
//   empirical actual
//
// This seems to show that, at least for x86_64, an accurate estimate is:
//
//   malloc_usable_size() + sizeof(void*)
//
// I don't know why we need to add sizeof one void* instead of two.
// I didn't try this with -m32 (32-bit model).
// (LarryRuane@gmail.com)
//
#include <malloc.h>
#include <algorithm>
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <unistd.h>

size_t heapsize()
{
    std::ifstream maps("/proc/" + std::to_string(getpid()) + "/maps");
    std::string line;
    while (std::getline(maps, line)) {
        if (line.find("[heap]") == std::string::npos) continue;
        // example:
        // 563579244000-56358c384000 rw-p 00000000 00:00 0 [heap]
        std::string::size_type dash_pos{line.find('-')};
        std::string::size_type space_pos{line.find(' ')};
        std::string start_str{line.substr(0, dash_pos)};
        std::string end_str{line.substr(dash_pos+1, space_pos-dash_pos-1)};

        size_t start{std::stoul(start_str, nullptr, 16)};
        size_t end{std::stoul(end_str, nullptr, 16)};
        return end - start;
    }
    return 0;
}

constexpr size_t n_allocations{100'000};
int main(int argc, char** argv) {
    for (size_t s{0}; s < 120; ++s) {
        size_t start{heapsize()};
        void* p;
        for (int i = 0; i < n_allocations; ++i)
            p = malloc(s);
        size_t end{heapsize()};
        double actual{double(end-start) / n_allocations};

        // empirically-derived, 64-bit linux x86_64
        size_t estimate{std::max<size_t>(32, ((s+7)/16+1)*16)};
        std::cout << s << ' '
            << malloc_usable_size(p) + sizeof(void*) << ' '
            << estimate << ' '
            << actual << '\n';
    }
    return 0;
}
```
</details>

**Comment by achow101** [comment]
Needs rebase if still relevant.

**Comment by LarryRuane** [comment]
This is still relevant, rebased, but changing to Draft because it should build on another PR I'm working on to implement https://github.com/bitcoin/bitcoin/pull/27748#discussion_r1209810958.

**Comment by martinus** [comment]
@LarryRuane are you still working on this? I think this shouldn't be blocked by #28531.

**Comment by martinus** [comment]
The PRs are very similar and achieve the same thing, with a bit different styles. Some differences:
* Here only system memory is continuously tracked and the rest is calculated on demand. In #28939 I track everything continuously. The result should be the same
* In #28939 I moved `MallocUsage` into a separate function to avoid code duplication
* I added another unit test in #28939 and fixed failing tests due to the new calculation (sometimes the old calculation understimated things)

**Comment by LarryRuane** [comment]
Your PR #28939 is definitely better, I especially like tracking total allocation continuously. I'll go ahead and close this.
